src/main/java/org/prebid/server/auction/BidResponseCreator.java

```diff
@@ -100,23 +114,137 @@ BidResponse create(List<BidderResponse> bidderResponses, BidRequest bidRequest,
                         winningBids, winningBidsByBidder, cacheInfo, cacheResult.getCacheBids(), account))
+                .map(seatBids -> bidResponseBuilder.seatbid(seatBids).build());
+    }
+
+    /**
+     * Corresponds cacheId (or null if not present) to each {@link Bid}.
+     */
+    private Future<CacheServiceResult> toBidsWithCacheIds(List<BidderResponse> bidderResponses, Set<Bid> bidsToCache,
+                                                          List<Imp> imps, BidRequestCacheInfo cacheInfo,
+                                                          Account account, Timeout timeout) {
+        final Future<CacheServiceResult> result;
+
+        if (!cacheInfo.isDoCaching()) {
+            result = Future.succeededFuture(CacheServiceResult.of(null, null, toMapBidsWithEmptyCacheIds(bidsToCache)));
+        } else {
+            // do not submit bids with zero price to prebid cache
+            final List<Bid> bidsWithNonZeroPrice = bidsToCache.stream()
+                    .filter(bid -> bid.getPrice().compareTo(BigDecimal.ZERO) > 0)
+                    .collect(Collectors.toList());
+
+            final boolean shouldCacheVideoBids = cacheInfo.isShouldCacheVideoBids();
+            final boolean eventsEnabled = Objects.equals(account.getEventsEnabled(), true);
+
+            final List<String> videoBidIdsToModify = shouldCacheVideoBids && eventsEnabled
+                    ? getVideoBidIdsToModify(bidderResponses, imps)
+                    : Collections.emptyList();
+
+            final CacheContext cacheContext = CacheContext.builder()
+                    .cacheBidsTtl(cacheInfo.getCacheBidsTtl())
+                    .cacheVideoBidsTtl(cacheInfo.getCacheVideoBidsTtl())
+                    .shouldCacheBids(cacheInfo.isShouldCacheBids())
+                    .shouldCacheVideoBids(shouldCacheVideoBids)
+                    .videoBidIdsToModify(videoBidIdsToModify)
+                    .build();
+
+            result = cacheService.cacheBidsOpenrtb(bidsWithNonZeroPrice, imps, cacheContext, account, timeout)
+                    .map(cacheResult -> addNotCachedBids(cacheResult, bidsToCache));
+        }
+
+        return result;
+    }
+
+    /**
+     * Resolves what {@link Bid}s should be cached.
+     * <p>
+     * If {@link BidRequestCacheInfo#isShouldCacheWinningBidsOnly()} is true - group ImpIds to corresponding Bid
+     * with max price (e.i. winningBid) and return all winning bids;
+     * <p>
+     * Otherwise - return all bids as usual.
+     */
+    private static Set<Bid> getBids(List<BidderResponse> bidderResponses, boolean shouldCacheWinningBidsOnly) {
+        final Set<Bid> bids = bidderResponses.stream()
+                .flatMap(BidResponseCreator::getBids)
+                .collect(Collectors.toSet());
+
+        if (!shouldCacheWinningBidsOnly) {
+            return bids;
+        }
+
+        final HashSet<Bid> bids1 = new HashSet<>(bids.stream()
```

Comment:
Looks like a draft ;)

Reply:
The whole method was a draft left after refactoring - it's not used anywhere. Good catch, my bad.